lib/Conversion/TritonGPUToLLVM/ScanOpToLLVM.cpp

```diff
@@ -0,0 +1,308 @@
+  if (!acc) {
+    acc = cur;
+    return;
+  }
+  // Create a new copy of the reduce block, and inline it
+  Block *currentBlock = rewriter.getBlock();
+  Region &parent = *currentBlock->getParent();
+  rewriter.cloneRegionBefore(combineOp, &parent.front());
+  auto &newScan = parent.front();
+  auto returnOp = dyn_cast<triton::ScanReturnOp>(newScan.getTerminator());
+  llvm::SmallVector<Value> combineArgs = {acc, cur};
+  rewriter.inlineBlockBefore(&newScan, &*rewriter.getInsertionPoint(),
+                             combineArgs);
+  auto results = returnOp.getResult();
+  acc = results[0];
+  // Delete the terminator, which is no longer used
+  rewriter.eraseOp(returnOp);
+}
+
+// Scan a contiguous elements within a thread and update `srcValues` in place.
+static void scanThreadContiguousElements(SmallVector<Value> &srcValues,
+                                         ConversionPatternRewriter &rewriter,
+                                         ScanLoweringHelper &helper) {
+  // TODO: this assumes that axis is the fastest moving dimension. We should
+  // relax that.
+  unsigned scanElementsPerThreads = helper.getAxisNumElementsPerThreads();
+  // Loop through the blocks of contiguous elements.
+  for (unsigned j = 0; j < srcValues.size(); j += scanElementsPerThreads) {
+    // Reset the accumulator at the beginning of each block of contiguous
+    // elements.
+    Value acc;
+    // Loop through the contiguous elements.
+    for (unsigned i = 0; i < scanElementsPerThreads; ++i) {
+      accumulate(rewriter, helper.getCombineOp(), acc, srcValues[i + j]);
+      srcValues[i + j] = acc;
+    }
+  }
+}
+
+// Apply a scan across threads of the warp for the last element of each
+// contiguous group of elements.
+static void warpScan(SmallVector<Value> &srcValues,
+                     ConversionPatternRewriter &rewriter,
+                     ScanLoweringHelper &helper, Value laneId) {
+  Location loc = helper.getLoc();
+  unsigned scanElementsPerThreads = helper.getAxisNumElementsPerThreads();
+  for (unsigned j = scanElementsPerThreads - 1; j < srcValues.size();
+       j += scanElementsPerThreads) {
+    Value acc = srcValues[j];
+    unsigned scanDim = helper.getAxisNumThreadsPerWarp();
+    // Reduce within warps.
+    for (unsigned i = 1; i <= scanDim / 2; i = i << 1) {
+      Value shfl = shflUpSync(loc, rewriter, acc, i);
+      Value tempAcc = acc;
+      accumulate(rewriter, helper.getCombineOp(), tempAcc, shfl);
+      Value mask = icmp_slt(laneId, i32_val(i));
+      acc = select(mask, acc, tempAcc);
+    }
+    srcValues[j] = acc;
+  }
+}
+
+// For each set of contiguous elements within a thread we store the partial
+// reduction into shared memory. Each parallel scan and each warp will store its
+// own partial reductions. The shared memory is organized as follow:
+//          -----------------------------------------------------------------
+// chunk 0: | scan 0 warp 0 | scan 1 warp 0 | scan 0 warp 1 | scan 1 warp 1 |
```

Comment:
It's not clear to me what scan 1 and scan 0 are.
I get the idea though after reading the code

Reply:
those numbers are meant to be the non-axis dimension. I improved the comment a bit. Let me know if you think it could still be clarified.